### ISSUE TITLE
server wide user listing

### ISSUE THREAD

**Issue by ddumontatibm** [mfilosa/Monitor-Docs#4]
I don't recall if you said this was coming or already available?

Is there a way to get the whole list instead of having it paged at x most recent?   I'll need the whole list each time.

**Comment by beetle-ibm** [comment]
@ddumontatibm as I understand it there is one api call that returns ALL users. (GET /users) It has many fields but not email.  There is another API call to get the details on a specific user (GET /users/:username) which does include email.  Currently I think we have to call the first one and the iterate through each user name.  

https://developer.github.com/v3/users/

**Comment by ddumontatibm** [comment]
I'd rather not be making 240 xhr requests every day to fill out this log..    I don't want to have to deal with the network issues that might crop up.

**Comment by beetle-ibm** [comment]
Agreed.  The get all users call already returns a lot of info... I see no reason it should not return emails as well, and one would think it should not be a hard change for github to make.   That said - I am not sure if that will happen in the next 45 days and in the meantime I can't imagine iterating through even 1000 user calls each night, each returning perhaps a kb of info would cause any problems with network or server load.

**Comment by mfilosa** [comment]
@beetle-ibm @ddumontatibm Iterating over he list is the best path forward for this I would not expect this to have any impact on server load. 

Iterate over get all users ( https://developer.github.com/v3/users/#get-all-users ), using list all email addresses on each ( https://developer.github.com/v3/users/emails/#list-email-addresses-for-a-user )

There is a user report that contains the email but this I don't believe can be pulled via an API. It can be run from the server.  The user report should have their primary e-mail address. You can access this via 
admin pages (rocket icon in the top right) -> reports

**Comment by ddumontatibm** [comment]
Sorry @mfilosa this current solution *does not* scale.   Whitewater cannot have the logging framework make thousands of requests, 1 at a time, for email addresses for all users in it.  We need this in the user api return.

**Comment by ddumontatibm** [comment]
I can get away with it for our small install, but the larger issue remains.

**Comment by mfilosa** [comment]
ok @ddumontatibm thanks, @celsmore lets make sure we make this part of the feature requests we add to the feature request.

@ddumontatibm let be be sure on what you need. 

You need a list of all users which contains their email address in the response for each correct. 

Adding email address to the response on this API (https://developer.github.com/v3/users/#get-all-users))  would satisfy this but I want to make sure we capture the underlying requirement. Do I have this correct?

**Comment by ddumontatibm** [comment]
This looks like it will, yes.  

We need to triangulate the following data:  username, userid, email.   If they are all there, it's much simpler to use this api to grab the list of users.

**Comment by ddumontatibm** [comment]
I ran into this https://help.github.com/enterprise/2.1/admin/articles/command-line-utilities/#ghe-user-csv
which seems perfect!

**Comment by mfilosa** [comment]
@ddumontatibm I cant believe we missed that. 

It likely returns the primary email it is possible there could be multiple.

**Comment by ddumontatibm** [comment]
If you could confirm that, I think that would still meet our needs.

**Comment by beetle-ibm** [comment]
to confirm... are user numeric ID's unique forever?  Like if a user with numeric id 10 is deleted the next user to be created isn't given 10 as their ID.... is that correct?

**Comment by beetle-ibm** [comment]
Also is it safe to assume there can't be two users with the same name but different cases?  Like "user" and "User"

**Comment by beetle-ibm** [comment]
I just realized... that csv user output does not seem to contain numeric ID... how hard would it be to add it?  Maybe we can get away without it... so far it looks like we have username everywhere... but I would hate to find some place I need the numeric ID and not be able to have it...

**Comment by ddumontatibm** [comment]
@beetle-ibm I do not know what the rules are for ldap for the uid.   The uid is unique but I'm not sure if it's case sensitive or not.

It wouldn't be bad to assume they *are* case sensitive, would it?

**Comment by beetle-ibm** [comment]
@ddumontatibm Yes - it could be bad to assume that they *are* case sensitive... If I assume that and they really are not case sensitive then how can I trust that I will always see them in the same case in all places?  I see some capitals in the user file... that doesn't mean there is definitely no place in the logs where the username is lower (or upper) cased.    Github does not seem to be very consistent in logging... As mfilosa said.... 

>We don't have normalized log output from all the various services that are shipping logs to audit.log

For instance in the audit log, some rows call repo "repo" other rows call it "repo_name".  Some rows call user name "user" some rows call user name "user_login".  If I just assume they are case sensitive and they are not, and some part of github is lower casing, then I might think they are different users when they are not.

**Comment by ddumontatibm** [comment]
ok, but they would have to go out of their way to change the case in what they log.   The username in our case comes from ldap and is not editable by the user.

**Comment by mfilosa** [comment]
> If you could confirm that, I think that would still meet our needs.

@ddumontatibm I have tested this myself and in my tests the report includes the primary email. If a user has a second email address in github it will not be included in the report. 

> I just realized... that csv user output does not seem to contain numeric ID... how hard would it be to add it? Maybe we can get away without it... so far it looks like we have username everywhere... but I would hate to find some place I need the numeric ID and not be able to have it...

I have some good news on this front. The userid does appear in the report generated from the Admin Dashboard --> Reports section of GitHub Enterprise. This report can be programmatically generate as well as generated in the UI.

To generate this programmatically you can use something like curl to do a get to the server. As an example:

`` curl -L -u username:password http(s)://hostname/stafftools/reports/all_users.csv ``

This syntax should work for the other reports as well. More information on dashboard reports can be found here: https://help.github.com/enterprise/2.3/admin/articles/site-admin-dashboard/#reports

I am hoping this will resolve this thread so I am closing this issue.

**Comment by ddumontatibm** [comment]
Will this work with an access token?   I hope you understand the complexity involved in having to have functional IDs to access this kind of stuff.   The ssh key was ideal, now with another set of creds to manage, this complicates things. 

Any chance of getting id numbers in the command line tool?

**Comment by ddumontatibm** [comment]
user account passwords rotate every 90 days.
